tests/Doctrine/Tests/Common/Proxy/AbstractProxyFactoryTest.php

```diff
@@ -215,6 +216,170 @@ public function testMissingPrimaryKeyValue()
+        $definition      = new ProxyDefinition('MyObject1', [], [], null, null);
+        $proxyGenerator  = $this->createMock(ProxyGenerator::class);
+        $metadataFactory = $this->createMock(ClassMetadataFactory::class);
+
+        $metadataFactory
+            ->expects($this->once())
+            ->method('getMetadataFor')
+            ->willReturn($metadata);
+
+        $metadata
+            ->expects($this->once())
+            ->method('getName')
+            ->willReturn('MyObject1');
+
+        $proxyGenerator
+            ->expects($this->once())
+            ->method('getProxyFileName')
+            ->willReturn($proxyFile);
+
+        $proxyGenerator
+            ->expects($this->once())
+            ->method('generateProxyClass')
+            ->willReturnCallback(function() use ($proxyFile) {
+                file_put_contents($proxyFile, '<?php class MyObject1 {} ');
+            });
+
+        /** @var MockObject&AbstractProxyFactory $proxyFactory */
+        $proxyFactory = $this->getMockForAbstractClass(
+            AbstractProxyFactory::class,
+            [$proxyGenerator, $metadataFactory, AbstractProxyFactory::AUTOGENERATE_FILE_NOT_EXISTS_OR_CHANGED]
+        );
+
+        $proxyFactory
+            ->expects($this->any())
+            ->method('createProxyDefinition')
+            ->will($this->returnValue($definition));
+
+        $generatedProxy = $proxyFactory->getProxy('Class', ['id' => 1]);
+
+        self::assertInstanceOf('MyObject1', $generatedProxy);
+    }
+
+    public function testGetProxyFileWhenProxyIsOlderThanSource()
+    {
+        $proxyFile = tempnam(sys_get_temp_dir(), 'proxy');
+        file_put_contents($proxyFile, '<?php class MyObject2 {} ');
+        sleep(1);
+        $sourceFile = tempnam(sys_get_temp_dir(), 'source');
+
+        $metadata        = $this->createMock(ClassMetadata::class);
+        $definition      = new ProxyDefinition('MyObject2', [], [], null, null);
+        $proxyGenerator  = $this->createMock(ProxyGenerator::class);
+        $metadataFactory = $this->createMock(ClassMetadataFactory::class);
+        $reflection      = $this->createMock(ReflectionClass::class);
+
+        $metadataFactory
+            ->expects($this->once())
+            ->method('getMetadataFor')
+            ->willReturn($metadata);
+
+        $metadata
+            ->expects($this->once())
+            ->method('getName')
+            ->willReturn('MyObject2');
+
+        $metadata
+            ->expects($this->once())
+            ->method('getReflectionClass')
+            ->willReturn($reflection);
+
+        $reflection
+            ->expects($this->once())
+            ->method('getFileName')
+            ->willReturn($sourceFile);
+
+        $proxyGenerator
+            ->expects($this->once())
+            ->method('getProxyFileName')
+            ->willReturn($proxyFile);
+
+        $proxyGenerator
+            ->expects($this->once())
+            ->method('generateProxyClass');
```

Comment:
Lots of mocking above. Can you please look around for real implementations that you could use instead? It make the test longer than it needs to be IMO. Also, I think the calls to `expects` can be dropped, because that's not really what you want to test here. That would make the above objects stubs instead of mocks, so you may try using `createStub` instead of `createMock` here (although I don't guarantee it will work, that depends on whether all builds use a recent enough version of PHPUnit)

Reply:
Yeah, just looked into it and everything is either an interface or not an easy to instantiate class, so I'm afraid you won't be able to reduce it more :disappointed: 
Functional tests would probably be better here, but that's a topic that is orthogonal to your PR.